File: apps/faithform-android/app/src/main/kotlin/io/faithform/app/ui/groups/GroupEdit.kt
package io.faithform.app.ui.groups

import androidx.compose.foundation.layout.*
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.input.KeyboardType
import io.faithform.app.contract.GroupDetail
import kotlinx.coroutines.launch
import kotlinx.serialization.json.*

@Composable fun GroupEdit(store: GroupsStore, detail: GroupDetail, onDismiss: () -> Unit, saved: () -> Unit) {
    var name by remember { mutableStateOf(detail.group.name) }; var description by remember { mutableStateOf(detail.description ?: "") }
    var enrollment by remember { mutableStateOf(detail.group.enrollment) }; var capacity by remember { mutableStateOf(detail.group.capacity?.toString() ?: "") }
    var location by remember { mutableStateOf(detail.location?.name ?: "") }; var address by remember { mutableStateOf(detail.location?.address ?: "") }; var url by remember { mutableStateOf(detail.location?.onlineMeetingUrl ?: "") }
    var posting by remember { mutableStateOf(detail.chatPosting ?: detail.group.chat?.postingPolicy ?: "everyone") }; var roster by remember { mutableStateOf(detail.memberListVisibility ?: "members") }; val scope = rememberCoroutineScope()
    GroupFormSheet("Edit group", onDismiss) {
        OutlinedTextField(name, { name = it.take(80) }, label = { Text("Group name") }, modifier = Modifier.fillMaxWidth())
        OutlinedTextField(description, { description = it.take(4000) }, label = { Text("What’s your group about?") }, minLines = 3, modifier = Modifier.fillMaxWidth())
        GroupChoice("Who can join?", enrollment, listOf("open" to "Anyone can join", "approval_required" to "Leader approval", "invitation_only" to "By invitation", "closed" to "Closed for now")) { enrollment = it }
        OutlinedTextField(capacity, { capacity = it.take(6) }, label = { Text("Member limit (optional)") }, keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Number), modifier = Modifier.fillMaxWidth())
        OutlinedTextField(location, { location = it.take(200) }, label = { Text("Meeting place") }, modifier = Modifier.fillMaxWidth())
        OutlinedTextField(address, { address = it.take(500) }, label = { Text("Address") }, modifier = Modifier.fillMaxWidth())
        OutlinedTextField(url, { url = it.take(2000) }, label = { Text("Online meeting link") }, keyboardOptions = KeyboardOptions(keyboardType = KeyboardType.Uri), modifier = Modifier.fillMaxWidth())
        GroupChoice("Who can post?", posting, listOf("everyone" to "All members", "leaders" to "Leaders only")) { posting = it }
        GroupChoice("Who can see the member list?", roster, listOf("members" to "All members", "leaders" to "Leaders only")) { roster = it }
        GroupFeedback(store)
        Button(enabled = !store.busy && name.isNotBlank(), onClick = { scope.launch {
            if (capacity.isNotBlank() && (capacity.toIntOrNull() ?: 0) <= 0) { store.error = "Enter a valid member limit or leave it blank."; return@launch }
            if (store.action("Group details saved.") { store.send<GroupDetail>("${store.path}/${detail.group.id}", buildJsonObject {
                put("expectedVersion", detail.group.version); put("name", name); put("description", description); put("enrollment", enrollment); put("capacity", capacity.toIntOrNull()?.let(::JsonPrimitive) ?: JsonNull)
                put("locationName", location); put("locationAddress", address); put("onlineMeetingUrl", url); put("chatPosting", posting); put("memberListVisibility", roster)
            }, "PATCH") }) { saved(); onDismiss() }
        } }, modifier = Modifier.fillMaxWidth()) { Text(if (store.busy) "Saving…" else "Save changes") }
    }
}
@Composable private fun GroupChoice(title: String, value: String, options: List<Pair<String, String>>, select: (String) -> Unit) {
    var expanded by remember { mutableStateOf(false) }
    Column { Text(title, style = MaterialTheme.typography.labelMedium); Box { OutlinedButton(onClick = { expanded = true }, modifier = Modifier.fillMaxWidth()) { Text(options.firstOrNull { it.first == value }?.second ?: value) }; DropdownMenu(expanded, { expanded = false }) { options.forEach { (key, label) -> DropdownMenuItem(text = { Text(label) }, onClick = { select(key); expanded = false }) } } } }
}

File: apps/faithform-ios/App/Groups/GroupEdit.swift
import SwiftUI
import FaithFormKit

struct GroupEditView: View {
    @Bindable var model: GroupsModel
    let detail: GroupDetail
    let saved: () -> Void
    @Environment(\.dismiss) private var dismiss
    @State private var name: String
    @State private var description: String
    @State private var enrollment: String
    @State private var capacity: String
    @State private var location: String
    @State private var address: String
    @State private var meetingUrl: String
    @State private var posting: String
    @State private var roster: String
    @State private var photoURL: String?
    @State private var version: Int
    init(model: GroupsModel, detail: GroupDetail, saved: @escaping () -> Void) {
        _photoURL = State(initialValue: detail.group.coverImageUrl)
        _version = State(initialValue: detail.group.version)
        self.model = model; self.detail = detail; self.saved = saved
        _name = State(initialValue: detail.group.name); _description = State(initialValue: detail.description ?? "")
        _enrollment = State(initialValue: detail.group.enrollment); _capacity = State(initialValue: detail.group.capacity.map(String.init) ?? "")
        _location = State(initialValue: detail.location?.name ?? ""); _address = State(initialValue: detail.location?.address ?? "")
        _meetingUrl = State(initialValue: detail.location?.onlineMeetingUrl ?? ""); _posting = State(initialValue: detail.chatPosting ?? detail.group.chat?.postingPolicy ?? "everyone")
        _roster = State(initialValue: detail.memberListVisibility ?? "members")
    }
    var body: some View {
        NavigationStack { Form {
            Section("Group photo") {
                GroupCoverView(url: photoURL, name: name).frame(height: 160).clipped()
                BrandingPhotoControl(title: "group photo", aspect: 16 / 9, hasPhoto: photoURL != nil) { data in
                    await model.perform("Group photo saved.") {
                        let result = try await model.send("\(model.path)/\(detail.group.id)/photo", method: .put, body: BrandingPhotoBody(imageBase64: data?.base64EncodedString()), as: BrandingPhotoResult.self)
                        photoURL = result.url
                        let latest = try await model.read("\(model.path)/\(detail.group.id)", as: GroupDetail.self)
                        version = latest.group.version
                        saved()
                        await model.load()
                    }
                }.disabled(model.busy)
            }
            Section("Make people feel welcome") { TextField("Group name", text: $name); TextField("What’s your group about?", text: $description, axis: .vertical).lineLimit(3...8) }
            Section("Joining your group") {
                Picker("Who can join?", selection: $enrollment) { Text("Anyone can join").tag("open"); Text("Leader approval").tag("approval_required"); Text("By invitation").tag("invitation_only"); Text("Closed for now").tag("closed") }
                TextField("Member limit (leave blank for unlimited)", text: $capacity).keyboardType(.numberPad)
            }
            Section("When you get together") { TextField("Meeting place", text: $location); TextField("Address", text: $address); TextField("Online meeting link", text: $meetingUrl).keyboardType(.URL).textInputAutocapitalization(.never).autocorrectionDisabled() }
            Section("Conversation & privacy") {
                Picker("Who can post?", selection: $posting) { Text("All members").tag("everyone"); Text("Leaders only").tag("leaders") }
                Picker("Who can see the member list?", selection: $roster) { Text("All members").tag("members"); Text("Leaders only").tag("leaders") }
            }
            GroupFeedback(model: model)
        }.navigationTitle("Edit group").navigationBarTitleDisplayMode(.inline).toolbar {
            ToolbarItem(placement: .cancellationAction) { Button("Cancel") { dismiss() } }
            ToolbarItem(placement: .confirmationAction) { Button(model.busy ? "Saving…" : "Save") { Task {
                guard capacity.isEmpty || (Int(capacity).map { $0 > 0 } == true) else { model.error = "Enter a valid member limit or leave it blank."; return }
                if await model.perform("Group details saved.", operation: {
                    _ = try await model.send("\(model.path)/\(detail.group.id)", method: .patch, body: GroupEditBody(value: UpdateGroupDetailsRequest(expectedVersion: version, name: name, description: description, enrollment: enrollment, capacity: Int(capacity), locationName: location, locationAddress: address, onlineMeetingUrl: meetingUrl, chatPosting: posting, memberListVisibility: roster)), as: GroupDetail.self)
                }) { saved(); dismiss() }
            } }.disabled(model.busy || name.trimmingCharacters(in: .whitespaces).isEmpty) }
        } }
    }
}

private struct GroupEditBody: Encodable, Sendable {
    let value: UpdateGroupDetailsRequest
    enum CodingKeys: String, CodingKey { case expectedVersion, name, description, enrollment, capacity, locationName, locationAddress, onlineMeetingUrl, chatPosting, memberListVisibility }
    func encode(to encoder: any Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(value.expectedVersion, forKey: .expectedVersion); try c.encode(value.name, forKey: .name)
        try c.encode(value.description, forKey: .description); try c.encode(value.enrollment, forKey: .enrollment)
        try c.encode(value.capacity, forKey: .capacity); try c.encode(value.locationName, forKey: .locationName)
        try c.encode(value.locationAddress, forKey: .locationAddress); try c.encode(value.onlineMeetingUrl, forKey: .onlineMeetingUrl)
        try c.encode(value.chatPosting, forKey: .chatPosting); try c.encode(value.memberListVisibility, forKey: .memberListVisibility)
    }
}
